R-package/src/lightgbm_R.cpp

```diff
@@ -284,29 +309,37 @@ SEXP LGBM_DatasetGetField_R(SEXP handle,
     }
   }
   R_API_END();
+  UNPROTECT(1);
+  return R_NilValue;
 }
 
 SEXP LGBM_DatasetGetFieldSize_R(SEXP handle,
   SEXP field_name,
   SEXP out) {
-  R_API_BEGIN();
-  const char* name = CHAR(Rf_asChar(field_name));
+  const char* name = CHAR(PROTECT(Rf_asChar(field_name)));
   int out_len = 0;
   int out_type = 0;
   const void* res;
+  R_API_BEGIN();
   CHECK_CALL(LGBM_DatasetGetField(R_ExternalPtrAddr(handle), name, &out_len, &res, &out_type));
   if (!strcmp("group", name) || !strcmp("query", name)) {
     out_len -= 1;
   }
   INTEGER(out)[0] = out_len;
   R_API_END();
+  UNPROTECT(1);
+  return R_NilValue;
 }
 
 SEXP LGBM_DatasetUpdateParamChecking_R(SEXP old_params,
   SEXP new_params) {
+  const char* old_params_ptr = CHAR(PROTECT(Rf_asChar(old_params)));
+  const char* new_params_ptr = CHAR(PROTECT(Rf_asChar(new_params)));
   R_API_BEGIN();
-  CHECK_CALL(LGBM_DatasetUpdateParamChecking(CHAR(Rf_asChar(old_params)), CHAR(Rf_asChar(new_params))));
+  CHECK_CALL(LGBM_DatasetUpdateParamChecking(old_params_ptr, new_params_ptr));
   R_API_END();
+  UNPROTECT(2);
```

Comment:
In some places in this PR, you chose to increment a counter `n_protected` and then call `UNPROTECT(n_protected)`. In other places, you used code like this that just hard-codes a number.

Is there a reason for that? If not, I'd like to use a consistent pattern throughout this file. I have a mild preference for the pattern where you use an integer literal like `UNPROTECT(2)`, just for code conciseness. I think we could rely on a CI job soon to catch mistakes of the form "imbalance of UNPROTECT() and PROTECT() calls" (#4400).

Reply:
Good point. No particular reason... I thought about that myself and agree with your assessment. I will change this.